### ISSUE TITLE
send information about detection

### ISSUE THREAD

**Issue by AshleyRoth** [AlexeyAB/darknet#605]
Some time ago I was already creating a topic about sending data about recognized objects. Unfortunately, as such, I received no answers.
prompted where in the code darknet it is possible to address for reception of the information on recognition of a concrete class.

I wrote the client - server code. I run the server on the same computer. The client code inserted in image.c after:
`if (class_id> = 0) {int width = show_img-> height * .006;`

int width = show_img-> height * .006;

if (strcmp (names [class_id], "dog") == 0) {
printf ("detect a dog! \ n");
dog = 6; // here I assign a variable to the variable to check in the client's condition, and if it is, send a message to the server
send_v2 ();// here I call the client
}

Unfortunately when I start detection on video, after the first detection and sending of information to the server, darknet freezes.

in the code of my client there is a loop, it seems to me that the problem in this

```
  while (1)
     {
     if (dog == 6) {
                 k = send (sock_desc, "dog", 100.0)
     }
     }
```

Can anyone tell me how to solve this problem for me? In order for darknet to continue its detection and not hang after the first condition? Please

**Comment by AshleyRoth** [comment]
@AlexeyAB I try, It did not help( But if i comment  `k=connect(sock_desc,(s...` it's not freezing. Oo Do you have any ideas what could be the problem? why can hang when connect there? connect I use to connect to the server

client.sin_family=AF_INET;
client.sin_addr.s_addr=inet_addr("127.0.0.1");
client.sin_port=1010;
```
  //  serv=connect(sock_desc,(struct sockaddr*)&client,sizeof(client)); 
  //  while(1)
  //  {
		if (dog==6) {
        k=send(sock_desc,"dog",100,0);

		} 
   // }
```

**Comment by AshleyRoth** [comment]
@AlexeyAB bwt can there be a problem with threads? For example, detection and client code can not work together

**Comment by AshleyRoth** [comment]
@AlexeyAB 
I need to update my comment after a little experiment. As a result, the whole problem lies in the loop. If the loop is removed, or after the condition the dog = 6, add a break, then there is no freezing. And I need to have a loop, in a loop there must be conditions for execution. How then can you solve this problem? Ideas? Thank you

**Comment by AlexeyAB** [comment]
@AshleyRoth Do you want to use Darknet as network-Server or as network-Client?

**Comment by AlexeyAB** [comment]
Add here: https://github.com/AlexeyAB/darknet/blob/0c95d8dfac7869ae4258f2080950cae6ea6a647e/src/image.c#L225


```cpp
if( strcmp(names[class_id], "dog") == 0)  // if "dog" is detected
{
    

    static int sock;
    static struct sockaddr_in server;
    static int initialized = 0;

    if(initialized == 0) {
        sock = socket(AF_INET , SOCK_STREAM , 0);

        server.sin_family = AF_INET;
        server.sin_addr.s_addr = inet_addr("127.0.0.1"); // server IP
        server.sin_port = htons( 1010 );  // server Port

       if (connect(sock , (struct sockaddr *)&server , sizeof(server)) == 0)
       {
          initialized = 1;  // successfully connected to the server
       }
    }


  if(initialized == 1) 
  {
    size_t length;
    char *ptr;
    ptr = (char*) names[class_id]; // send "dog"
    length = strlen(ptr);

    while (length > 0)
    {
        int i;
        i  = send(sock, ptr, length);
        if (i < 1) return false;
        ptr += i;
        length -= i;
    }
  }
}

```

**Comment by AshleyRoth** [comment]
@AlexeyAB , Thanks, i try it and back to you again :) But what about video? On the video, this code will work stably? My code as I already said in the while of freezes..

**Comment by AlexeyAB** [comment]
For video add the same code here in the function `draw_detections_cv_v3()` instead of `draw_detections_v3()`: https://github.com/AlexeyAB/darknet/blob/0c95d8dfac7869ae4258f2080950cae6ea6a647e/src/image.c#L363

**Comment by AshleyRoth** [comment]
@AlexeyAB on images are working, thanks. I check this on video and back again

Hmm, if I want to detect and send several classes, I will have to copy the condition several times. It seems to me so wrong already? If globally declare the connection and depending on the detection in the conditions to send to the server, how much is the correct solution to be?

**Comment by AlexeyAB** [comment]
Just change this line `if( strcmp(names[class_id], "dog") == 0) ` 
to this `if( strcmp(names[class_id], "dog") == 0 || strcmp(names[class_id], "cat") == 0) `

**Comment by AshleyRoth** [comment]
@AlexeyAB 
wah! it works, thanks a lot! Then why my code did not work, I can not understand it ...
There are two questions about your code:

1. when I do two classes, he connects the names. The output on the server looks like this: a dogcat. How do I make a dog on one line, cat on another? Example: 

> dog
> dog
> cat
> dog

2. If I understand correctly, he sends the information when he sees on each frame? Is it possible to implement it so that it detects the object-> sent dog-> zeroed. Then all over again. In my code, I tried to add the dog = 0 to the end of the condition IF
```
 //  while(1)
  //  {
		if (dog==6) {
        k=send(sock_desc,"dog",100,0);

		} 
   // }
```

**Comment by AlexeyAB** [comment]
1. Use this part of code:
```cpp
  if(initialized == 1) 
  {
    size_t length;
    char *ptr;
    ptr = (char*) names[class_id]; // send "dog"
    length = strlen(ptr);

    while (length > 0)
    {
        int i;
        i  = send(sock, ptr, length);
        if (i < 1) break;
        ptr += i;
        length -= i;
    }
    char * new_line = "\n";
    send(sock, new_line, strlen(new_line));
  }
```

2. I don't understand

**Comment by AshleyRoth** [comment]
@AlexeyAB

I meant when several objects are detected, then they connect them at the output. An example on the screenshot (unfortunately now it is necessary to check on the video from the car, so at the conclusion of knowledge, others).

2. Here he meant, he sends to the server all the time until he sees (without stopping). For example, there was a detection of the dog -> the message to the server was sent-> no longer sent (

**Comment by AlexeyAB** [comment]
Can you show whole code of the `draw_detections_v3()` functions that you use?

**Comment by AlexeyAB** [comment]
Instead of these lines:
```cpp
		  if(initialized == 1) 
		  {
		    size_t length;
		    char *ptr;
		    ptr = (char*) names[class_id]; // send class name
		    length = strlen(ptr);

		    while (length > 0)
		    {
			int i;
			i  = send(sock, ptr, length); //sending if while true
			if (i < 1) break;
			ptr += i;
			length -= i;
		    }
		  }
		}
```

Use these lines:
```cpp
		  if(initialized == 1) 
		  {
		    size_t length;
		    char *ptr;
		    ptr = (char*) names[class_id]; // send class name
		    length = strlen(ptr);

		    while (length > 0)
		    {
			int i;
			i  = send(sock, ptr, length); //sending if while true
			if (i < 1) break;
			ptr += i;
			length -= i;
		    }
                    char * new_line = "\n";
                    send(sock, new_line, strlen(new_line));
		  }
		}


```

**Comment by AshleyRoth** [comment]
@AlexeyAB nothing has changed, an empty string appears and that's it. conclusion is the same confusing

> dog
> dog
> dog
> cat
> catcat
> bear
> catr

**Comment by AlexeyAB** [comment]
Yes, it can be fixed using these lines
```
                    char * new_line = "\n";
                    send(sock, new_line, strlen(new_line));
```

**Comment by AshleyRoth** [comment]
@AlexeyAB , Hello! Excuse me, i have a question. I found where to get the x, y, width and height of the detected object. How can I send these coordinates `printf("%f %f %f %f\n", b.x, b.y, b.w, b.h)` with the name of the recognized object? (for send information about detection i use previously code, i don't know how send x,y,w and h). Thank you
I wanna like:
`car 0.548250 0.499895 0.057039 0.098823`

**Comment by AshleyRoth** [comment]
@AlexeyAB Hi! 
Can you help me please? as in the new version of image.c make the output of messages in the "test" mode (recognition in pictures). In the old version, you said to put the code in line 225 (https://github.com/AlexeyAB/darknet/blob/0c95d8dfac7869ae4258f2080950cae6ea6a647e/src/image.c#L225)

Thanks!

**Comment by AlexeyAB** [comment]
@AshleyRoth Hi,

Add it here: https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/image.c#L310

**Comment by AshleyRoth** [comment]
@AlexeyAB what about "class_id" ? there is no class_id, what should I do in this case? 

Regards!

**Comment by AshleyRoth** [comment]
@AlexeyAB 
instead of "class_id", I in "draw_detections_v3" used "names [selected_detections [i] .best_class]". Now I encounter another problem, I do not connect anywhere with the detection, although with video in the "demo" mode everything is fine. Maybe in the case of "test" a little differently done?

Regards

Thank you!

**Comment by AlexeyAB** [comment]
@AshleyRoth 


Yes, you can `selected_detections[i].best_class` and `names[selected_detections[i].best_class]`

>  I do not connect anywhere with the detection

What do you mean?

**Comment by AshleyRoth** [comment]
@AlexeyAB 
Once again, forgive the question. I'll try to give you a more extended answer. I have a separate function "send_detect_objects", this function is called in "detect_in_thread" in the file "demo.c".

1. Do I understand correctly - everything that relates to the "demo" mode is in the file "demo.c"?
2. Where, in that case, is what relates to the "test" mode? in the "detector.c"? 3. Where do I then call my "send_detect_object" function?

Ask if there are questions, so I can better give you an answer. I want to solve the current problem.

Thank you!

**Comment by AlexeyAB** [comment]
1. Yes, for Video - whole demo.c file, and this function in the image.c file: https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/image.c#L440

----

2. For Images - this function in the detector.c https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/detector.c#L1083-L1084
    and this function in the image.c https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/image.c#L278


-----


For Video you can use your function `send_detect_object()`
* or here - and create your own for-loop: https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/demo.c#L115

* or here - inside this loop: https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/image.c#L465

**Comment by AshleyRoth** [comment]
@AlexeyAB

Thanks for the detailed answer. I'm particularly interested in the case with pictures.
hence I have to send my "send_detect_object" function into the "detector.c" file in the "test_detector" function?

**Comment by AlexeyAB** [comment]
Yes, you can add here: https://github.com/AlexeyAB/darknet/blob/18d5e4f39c1441f2c21043ac3204b5cb279f8758/src/detector.c#L1148
the loop like this:
```cpp
            int i;
            for (i = 0; i < nboxes; ++i) {
                char buff[1024];
                int class_id = -1;
                float prob = 0;
                for (j = 0; j < l.classes; ++j) {
                    if (dets[i].prob[j] > thresh && dets[i].prob[j] > prob) {
                        prob = dets[i].prob[j];
                        class_id = j;
                    }
                }
                if (class_id >= 0) {
                    sprintf(buff, "%d %2.4f %2.4f %2.4f %2.4f\n", class_id, dets[i].bbox.x, dets[i].bbox.y, dets[i].bbox.w, dets[i].bbox.h);
                    send_detect_object(buff, sizeof(char), strlen(buff));
                }
            }
```

**Comment by AshleyRoth** [comment]
@AlexeyAB 
Thank you very much, I understood. You have expanded my knowledge! An incomprehensible moment, if in "demo" mode with "http_port" we can do stream video. Is there anything in the "test" mode? Or maybe "http_port" works for both cases? Once again very grateful for your answers!

**Comment by AlexeyAB** [comment]
@AshleyRoth `http_port` works only for Video (only for ./darknet detector demo ...).

**Comment by AshleyRoth** [comment]
@AlexeyAB
and there is no solution for working with pictures? (test mode)

**Comment by AshleyRoth** [comment]
@AlexeyAB, Thanks for your answers!

**Comment by AshleyRoth** [comment]
Hello again @AlexeyAB

could you help with this to me? in the case of "detector test", how will it look? Thank you

if (class_id == 14 && dets[i].prob[class_id] < 0.6) continue;

I think something like: 

if(selected_detections[i].best_class == 14) && ??? <0.6) continue;

**Comment by AlexeyAB** [comment]
@AshleyRoth 
Try to use this

`if(selected_detections[i].best_class == 14 && selected_detections[i].det.prob[selected_detections[i].best_class] <0.6) continue;`
